audio: base_fw: add platform layer to hw_config data

kv2019i opened the pull request:
First patch to address https://github.com/thesofproject/sof/issues/8391 - more to come later.

Some of the HW_CONFIG fields are platform specific and cannot be filled with generic means in generic code.

Handle this data via a simple platform abstraction to hook platform specific code to HW_CONFIG handling. Move out existing code for Intel cAVS and ACE platforms.

kv2019i commented:
@dbaluta @iuliana-prodan this might affect NXP targets for IPC4 not yet in main (but in long term make it easier for you). I'm planning to move more plat spescific data out from base_fw.c but wanted to share this earlier to get feedback on the approach.

lgirdwood reviewed (approved):
LGTM, but we long term we should be removing the platform specific files as Zephyr will abstract platform. This looks IPC specific.

kv2019i commented:
@lgirdwood wrote:
> LGTM, but we long term we should be removing the platform specific files as Zephyr will abstract platform. This looks IPC specific.

Ack, we are trackign this is https://github.com/thesofproject/sof/issues/7248. To me it seems we have a layer of data that is above the RTOS layer and is different between vendors. Currently we use the "platform" layer for this in SOF. You pass a "platform name" to build script Personally I think we need the ability to vary the build contents. Some (most) of this is in terms of Kconfig settings and DTS overrides, but e.g. for the IPC4 tuples, we need content that requires custom code (at least for now). The platform layer in SOF seems like a natural fit.

kv2019i commented:
Issue in CI on one of the Intel platforms, tagging with DNM while debugging. Please review the concept still.

kv2019i commented:
V3:
- added similar change for FW_CONFIG (easier probably to review in one go as FW_CONFIG has more vendor/platform specific bits)
- fixed a bug in V2
- remove unneeded ifdefs on IPC version in platform files

kv2019i commented:
V4:
- one more bugfix to V3 (sorry for the churn)

lyakh commented on src/audio/base_fw.c:
```diff
@@ -147,19 +148,11 @@ static int basefw_hw_config(uint32_t *data_offset, char *data)
-	tuple = tlv_next(tuple);
-	tlv_value_uint32_set(tuple, IPC4_LP_EBB_COUNT_HW_CFG, PLATFORM_LPSRAM_EBB_COUNT);
+	/* add platform specific tuples */
+	platform_basefw_hw_config(&plat_data_offset, (char *)tuple);
```
doesn't this break non-Intel platforms? Do you need a `weak` version? Or is this all Intel-only?

kv2019i replied:
@lyakh This is just like all functions in platform.h, all platforms need to provide one. Let me refresh the series with "posix" platform support, I missed that and that causes some of the CI checks to fail.

The one design open I had (and still have to some extent), is whether to make this a "bigger thing", e.g. something like a "platform/ipc-data.h". But I ended up not doing that and just putting thees in top-level platform.h as the longterm goal is to minimize platform interface (drop all platform/clk.h, platfrom/dma.h as these can all be handled via Zephyr interfaces, we just can't drop them yet as some platfroms still use XTOS). Especially as @lgirdwood is still not convinced we need platform.h at all (see https://github.com/thesofproject/sof/issues/7248).

One alternative would be just to add src/ipc/ipc4/basefw_intel.c and not try to abstract this now. I'd still want to keep base_fw.c clean of vendor specific parts, so it is easier to use IPC4 for all SOF targets.

lyakh reviewed (changes requested):
something is weird here... same commit twice?..

kv2019i commented:
@thesofproject/nxp Any thoughts on the high-level concept? So I guess the options would be:

1) this appraoch, extend platform to cover vendor specific bits of base_fw.c
2) not touch and add to platform.h, but add some new base_fw_intel.c 
3) keep current base_fw.c and incrementally move pieces of Intel-specific code to Zephyr when possible (I'm personally not convinced we can move all, @lgirdwood thinks we can move more)

I think some abraction is needed pretty quick to keep base_fw.c clean (e..g we have things like https://github.com/thesofproject/sof/pull/9031 coming all the time).

kv2019i commented:
V5:
- implement the new interface for "posix" platform

kv2019i commented:
V6: 
- fix build warning with SDK gcc version

lgirdwood commented:
> Changing to draft. It seems even with a fresh cycle of cleanup to the platform layer, there is still quite a bit of usage left in the tree, so adding these definitions to main platform.h will not scale. Also, I see many new PRs coming to base_fw.c, so I think the platform-specific layer will have to be grown later. IOW, we need to add some more explicit interface where to put platform specific basefw stuff (versus just putting them to platform.h).

Depends what you mean by platform specific code, SOF codebase needs to be generic and Zephyr will define the platform.

kv2019i commented:
V7:
- moved the new interface to a separate header (versus adding to main platform.h)
- add implementations for all IPC4 targets
- dropped the FW_CONFIG support for now, will add back once we have agreement on the approach

lgirdwood commented on src/platform/intel/ace/lib/base_fw_platform.c:
```diff
@@ -0,0 +1,32 @@
+	struct sof_tlv *tuple = (struct sof_tlv *)data;
+	uint32_t value;
+
+	tlv_value_uint32_set(tuple, IPC4_HP_EBB_COUNT_HW_CFG, PLATFORM_HPSRAM_EBB_COUNT);
```
We should put this in ```src/ipc/ipc4/intel.c``` and export several C functions that can be called where vendor data has to be installed (otherwise the are all ```static inline vendor_add_memory()``` and compile out).
The input data for ```IPC4_HP_EBB_COUNT_HW_CFG, PLATFORM_HPSRAM_EBB_COUNT``` has to come from DT, but this could be a secondary PR later on (we could have a comment saying this today).

kv2019i replied:
@lgirdwood That's what this patch now does, ipc4/intel.c code in src/platform/intel/ace/lib/base_fw_platform.c and generic stub in src/platform/posix/base_fw_platform.c. 

We can keep this simpler and put this in ipc4/intel.c, but that means we need to maintain a platform/vendor layer in base_fw.h/base_fw_vendor.h and have ifdefs in src/ipc. Current PR keeps the ifdefs out of main src/ipc and ipc/audio trees.

I'm ok with both and can update the PR. There is more code and data to move once the approach is agreed, hw_config is just a start).

kv2019i commented:
Finally figured out the mismatch between github workflow and my local build for fuzzer build, even did an OS upgrade to local build machine to match the github workflow. In the end the rootcause I was testing the pushed version locally, while the github flow was testing against newer SOF with https://github.com/thesofproject/sof/pull/8737 . That adds constructs to base_fw.h that no longer build with the posix target, and thus my 9012 fails.

kv2019i commented:
V8:
- fix the stub/fuzzer build when combined with #8737 
- may need rework if we go for simpler src/ipc/ipc4/inter.c (and not use src/platform for this)

kv2019i commented on zephyr/CMakeLists.txt:
```diff
@@ -183,6 +184,7 @@ if (CONFIG_SOC_SERIES_INTEL_ADSP_ACE)
 	# Platform sources
 	zephyr_library_sources(
 		${SOF_PLATFORM_PATH}/intel/ace/platform.c
+		${SOF_PLATFORM_PATH}/intel/ace/lib/base_fw_platform.c
```
@ujfalusi Ack. If I can tune this in follow-up? I anticipate the ace one to be the full implementation and thus I put the main implementation here under ace/lib. For cavs we can probably go with a minimal implementation even, to keep the Linux SOF driver happy, but just wanted some example for cavs here as well for this PR.

jsarha reviewed (approved):
LGTM

jsarha commented:
Just occurred to me, that on basefw_set side the only supported tlv type looks pretty Intel specific too. Does not happen to be this PR, but it would make sense to move that to platform specific source file too.

kv2019i commented:
@jsarha wrote:
> Just occurred to me, that on basefw_set side the only supported tlv type looks pretty Intel specific too. Does not happen to be this PR, but it would make sense to move that to platform specific source file too.

Ack, a good idea. Also, the function signature could be improved. I now just took the existing code as-is from basefw so it is easier to remove the new layout, but I think this could be improved as well. These functions are now passed a "char *" pointer with no knowledge of size and it's very unobvious how it is handled the the functions don't overwite allocated buffer for the IPC response.
